books/kestrel/c/syntax/types.lisp

```diff
@@ -39,111 +41,111 @@
-      and unknown types are always acceptable."))
-   (xdoc::p
-    "Besides the approximations noted above,
-     currently we do not capture atomic types [C17:6.2.5/20],
-     which we approximate as the underlying (argument) type.
-     We also do not capture @('typedef') names,
-     which we approximate as unknown types.
-     Furthermore, we do not capture qualified types [C17:6.2.5/26]."))
-  (:void ())
-  (:char ())
-  (:schar ())
-  (:uchar ())
-  (:sshort ())
-  (:ushort ())
-  (:sint ())
-  (:uint ())
-  (:slong ())
-  (:ulong ())
-  (:sllong ())
-  (:ullong ())
-  (:float ())
-  (:double ())
-  (:ldouble ())
-  (:floatc ())
-  (:doublec ())
-  (:ldoublec ())
-  (:bool ())
-  (:struct ((tag? ident-optionp)))
-  (:union ())
-  (:enum ())
-  (:array ())
-  (:pointer ())
-  (:function ())
-  (:unknown ())
-  :pred typep)
+(encapsulate ()
+  (set-induction-depth-limit 1)
+
+  (fty::deftagsum type
+    :short "Fixtype of C types [C17:6.2.5]."
+    :long
+    (xdoc::topstring
+     (xdoc::p
+      "Currently we do not model all the C types in detail,
+       but only an approximate version of them,
+       which still lets us perform some validation.
+       We plan to refine the types, and the rest of the validator,
+       to cover exactly all the validity checks prescribed by [C17]
+       (as well as applicable GCC extensions).")
+     (xdoc::p
+      "We capture the following types:")
+     (xdoc::ul
+      (xdoc::li
+       "The @('void') type [C17:6.2.5/19].")
+      (xdoc::li
+       "The plain @('char') type [C17:6.2.5/3].")
+      (xdoc::li
+       "The five standard signed integer types [C17:6.2.5/4]
+        and the corresponding unsigned integer types [C17:6.2.5/6].")
+      (xdoc::li
+       "The three real floating point types [C17:6.2.5/10].")
+      (xdoc::li
+       "The three complex types [C17:6.2.5/11].
+        These are a conditional feature,
+        but they must be included in this fixtype
+        because this fixtype consists of all the possible types.")
+      (xdoc::li
+       "The @('_Bool') type [C17:6.2.5/2].")
+      (xdoc::li
+       "A family of structure types [C17:6.2.5/20].
+        Structure types are characterized by an optional tag.
+        This is an approximation,
+        because there may be different structure types of a given tag,
+        or different tagless structure types.")
+      (xdoc::li
+       "A collective type for all union types [C17:6.2.5/20].
+        This is an approximation,
+        because there are different union types.")
+      (xdoc::li
+       "A collective type for all enumeration types [C17:6.2.5/20].
+        This is an approximation,
+        because there are different enumeration types.")
+      (xdoc::li
+       "A collective type for all array types [C17:6.2.5/20].
+        This is an approximation,
+        because there are different array types.")
+      (xdoc::li
+       "A parameterized pointer type [C17:6.2.5/20].
```

Comment:
I would rephrase this slightly to say something like "a pointer type that includes information about the pointed-to type", since "parameterized type" has often a more specific meaning in type theory (although C does not have those kind of parameterized types).